stack of LSTM

udibr opened the pull request:
LSTMstack acts as a plug-and-play replacement for a regular LSTM

rizar commented:
Thanks for sharing your code with us!

Unfortunately, this brick does not meet Blocks standard of modularity. We don't want to have `SimpleRecurrentStack`, `LSTMStack` and `GatedRecurrectStack` classes that share a lot of code. It should be possible to write a generic `RecurrentStack` brick, that accepts a prototype recurrent brick and build a stack of such bricks interleaved with `Fork`'s. There is even a ticket for that, #46 . Let me know if you are interested in implementing something like that, I would be very glad to help.

udibr commented:
I did implemented something like that at some point, but, every recurrent unit (LSTM, GRU, Linear) expects to see a different set of inputs (LSTM need both states and cells others just states)

As a result you cant put a generic stack in place of where you previously had an LSTM unit
(this is why I said "plug-and-play")

rizar commented:
I think in Blocks it is possible to make a generic stack. Given a recurrent brick, you can query its apply method what states it has and what sequential inputs it expects by simply `brick.apply.sequences` and `brick.apply.states`. By accessing those you can even write a stack that supports different components, e.g. one layer GRU and one layer LSTM.

udibr commented:
I have code for it and will try to check if it is working

On Mon, Jun 1, 2015 at 3:32 PM, Dmitry Bogdanov <notifications@github.com>
wrote:

> I think in Blocks it is possible to make a generic stack. Given a
> recurrent brick, you can query its apply method what states it has and what
> sequential inputs it expects by simply brick.apply.sequences and
> brick.apply.states. By accessing those you can even write a stack that
> supports different components, e.g. one layer GRU and one layer LSTM.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mila-udem/blocks/pull/688#issuecomment-107679175>.
>

udibr commented:
The code now is a general purpose stack of recurrent networks

rizar commented on blocks/bricks/recurrent.py:
```diff
@@ -630,3 +631,140 @@ def apply(self, *args, **kwargs):
+        self.children = self.transitions + self.forks
+
+        # Programmatically fix the parameters of the  recurrent decoration
+        setattr(self.apply, 'sequences', transitions[0].apply.sequences)
```
Cool, this is a new technique which I never though of.. and I like it!

The standard way to have application method properties defined at runtime was so far

```
@apply.property('inputs')
def apply_inputs(self):
    return self.wrapped.inputs
```

but seems like simply setting attributes of the `BoundApplication` also does the job!

rizar commented on blocks/bricks/recurrent.py:
```diff
@@ -630,3 +631,140 @@ def apply(self, *args, **kwargs):
+            transition.push_allocation_config()
+
+        for d, fork in enumerate(self.forks):
+            fork.input_dim = self.transitions[d].get_dim("states")
```
Hmm, I thought that all the states should be forked, but apparently it does not match what they do in the literature... What about forking the first states, i.e. `self.transition.apply.states[0]`?

udibr replied:
Using "states" is ugly but this specific name does not leave any chance for error. You cant same the same on states[0]. I can keep the "states" string as a class parameter that you can override. I can also generate a warning if "states" is not located in index 0 of self.transition.apply.states

rizar commented on blocks/bricks/recurrent.py:
```diff
@@ -630,3 +631,140 @@ def apply(self, *args, **kwargs):
+                    layer_kwargs["mask"] = kwargs.get("mask")
+
+            for t in ["states", "contexts", "outputs"]:
+                suffix = '_%d' % d
```
This formatting is not available in Python 3, please use `str.format` or simply `+` operator.

rizar commented:
This is awesome, and we definitely want it Blocks! I left some change proposals and I will leave some more a bit later, but in general I really like it.

udibr commented on blocks/bricks/recurrent.py:
```diff
@@ -630,3 +631,140 @@ def apply(self, *args, **kwargs):
+                layer_kwargs = dict(zip(
+                    self.normal_inputs(d),
+                    self.forks[d-1].apply(kwargs["states_%d" % (d-1)],
+                                          as_list=True)))
```
In this line I am using the states from the previous step in the sequence as the input to all of the layers (sync mode)
there is also the async option in which the output state of one layer is used as the input (through fork) of the next layer for the same step (lets call this async mode.) which one should I use?

rizar replied:
Just stick to Alex Graves papers, I believe they set de-facto standard. Seems like it means your `async` mode.

rizar commented on blocks/bricks/recurrent.py:
```diff
@@ -630,3 +631,168 @@ def apply(self, *args, **kwargs):
+        (layer) that is used, via fork, as the sequences (input) of the
+        next layer. The same element should also appear in the states
+        parameter of the apply method.
+    sync : bool
```
I would leave only async.

rizar commented on blocks/bricks/recurrent.py:
```diff
@@ -630,3 +631,168 @@ def apply(self, *args, **kwargs):
+    Build a stack of recurrent layers from a supplied list of
+    BaseRecurrent objects. Each object must have a `sequences`,
+    `contexts`, `states` and `outputs` parameters to its `apply` method.
+    It is assumed that the `sequences` of all objects has an "inputs"
```
I don't see such an assumption in the code.

rizar commented:
This implementation could be made significantly faster by having two pathways for `iterate=True` and `iterate=False`. If `iterate=True` one can first run the lower layer until the end of sequence and then fork all its states. This can save quite some time since GPU would be used more efficiently.

Let me know if you feel like implementing it, I could give some hints.

udibr commented:
I'm all for speed and will look into it. I have two questions:
1) How do I detect inside apply if iterate is true/false? or should I rewrite the recurrent wrapper for his?
2) I've noticed that the memory available on the GPU depend on the number of steps (time length of the sequence.) maybe it is faster than linear dependcy. Any idea why?

rizar commented:
1) I would do smth like 

```python
@recurrent
def _apply(self, ...):
     """Your current apply method."""
     pass
 
@application
def apply(self, *args, **kwargs):
    if kwargs.get('iterate', True):
         # faster implementation, which can however share some code with _apply
    else:
         return self._apply(*args, **kwargs)
```

`AttentionRecurrent` from `blocks/bricks/attention.py` can be an example of doing such things.

2) Growth is normal, but it definitely should be linear. If you use attention, there might be some memory issues.

udibr commented:
I think current code is OK for integration. If/when I will have a faster version I will send another pull request.
P.S. I now understand why memory is linear... In backprop you need to keep all the forward results so you can latter multiply them with the backward error signal...this is the only thing that needs to grow

dwf commented:
On Wed, Jun 3, 2015 at 2:57 PM, Dmitry Bogdanov <notifications@github.com>
wrote:

> 2) Growth is normal, but it definitely should be linear. If you use
> attention, there might be some memory issues.
>
@rizar Superlinear memory growth could be attributable to the optimizer
giving up earlier on larger graphs and not doing all the in-place stuff
that it could.

rizar commented:
@udibr , regarding the memory: exactly

Would it be acceptable for you to commit it to `blocks-extras` then? We have already quite a lot of code in Blocks which has certain deficiencies and really wants to be fixed, and it would be nice to keep the ready components and the one in development separated. Since Blocks is now a namespace package, if you have both `blocks` and `blocks-extras` installed on your machine, the same import paths work for both. That is when this class is moved to Blocks, you will not have to change any code that uses it, assuming that in `blocks-extras` it will be also in `blocks.bricks.recurrent` module.

rizar commented:
@dwf , forgive me my poor knowledge of Theano, but I don't understand why not in-place operations can cause super-linear memory usage growth.

udibr commented:
not sure where to put the code in blocks-extras?
blocks/bricks/recurrent.py - this will conflict with blocks file of the same name
so to create a new python file or to put it in
blocks/extras/bricks/recurrent.py
but then I will have to modify the code to do the import from blocks.extras.bricks.recurrent

rizar commented:
Ooops, seems line this doesn't work!

I guess we will merge this into master then. Sorry for long delays: I have both NIPS deadline and graduation in one week.

dwf commented:
@rizar: my train of thought/speculation goes like this:

* One way that Theano attempts to save memory is by doing operations
in-place when destructive operations are safe. It implements this by means
of a graph optimizer.
* Theano's graph optimizers are heuristic, as doing what we'd like it to do
in full generality is NP-complete. Also a heuristic: when to stop running
further passes of the optimizer.
* If the latter heuristic gives up earlier than it could or perhaps should
on a larger graph, some optimizations may not be done, and the graph of
larger size may not be as thoroughly optimized as a smaller graph of the
same general flavour.
* If some of the optimizations that are not done are memory-saving
optimizations (most notably in-place optimizations), one may see this
manifested in superlinear growth of memory usage even with what appears to
be a linear growth in the number of mathematical operations performed.

Does that make sense?

udibr commented:
I've added the fast mode you asked (now it is on by default) and also a bonus distribution of inputs to all layers. Implementing fast was done with very littly codeing :-) However, it consumes more memory so I had to reduce my block_size to fit everything into the GPU and as a result the speed improvement where small (I got 30% for 3 layers)

rizar commented:
Great!

`also a bonus distribution of inputs to all layers.` - isn't it typically called skip connections in literature, in particular [here](http://arxiv.org/abs/1308.0850)? 

30% is a good speedup, I don't know why you call it small. Can you give an estimate of how much the memory consumption increased?

rizar commented on blocks/bricks/recurrent.py:
```diff
@@ -630,3 +631,231 @@ def apply(self, *args, **kwargs):
+
+    Build a stack of recurrent layers from a supplied list of
+    BaseRecurrent objects. Each object must have a `sequences`,
+    `contexts`, `states` and `outputs` parameters to its `apply` method.
```
These are requirements from the `recurrent` decorator, so I would add `such as the ones required by the recurrent decorator from :mod:`blocks.bricks.recurrent`.

udibr replied:
Can you please have a look at the code for recurrent decorator? Perhaps it has an error in its docstring saying inputs and not sequences  and if so this should be fixed. However, if you look at the code when handling return_initial_states case there is a reference to .owner.inputs (I guess this is OK because return_initial_states is in use in SequenceGenerator)

rizar replied:
Thanks, this was actually a docstring error, see #702 . `.owner.inputs` should however be not confused with Blocks application method properties, these are inputs of a Theano apply node.

rizar commented on blocks/bricks/recurrent.py:
```diff
@@ -630,3 +631,231 @@ def apply(self, *args, **kwargs):
+    (this can be configured with `states_name` parameter.)
+
+    The `sequences` (the "inputs") of the the first layer (layer 0) is
+    used as the `sequence` of the stack (the instance of this class.)
```
typo: `sequences`

rizar commented on blocks/bricks/recurrent.py:
```diff
@@ -630,3 +631,231 @@ def apply(self, *args, **kwargs):
+    transitions : list
+        List of recurrent units to use in each layer.
+        Note: A suffix with layer number is added to transitions' names.
+    prototype : :class:`.FeedForward`, optional
```
Can you please rename it to `fork_prototype`? Simply `prototype` is a bit confusing.

rizar commented on blocks/bricks/recurrent.py:
```diff
@@ -630,3 +631,231 @@ def apply(self, *args, **kwargs):
+         :class:`.Linear` transformation with biases is used.
+    states_name : string
+        In a stack of RNN the state of each layer is used as input to the
+        next. This string identify which element in the output of each
```
`which state of each layer`, not `which element in the output`

rizar commented on blocks/bricks/recurrent.py:
```diff
@@ -630,3 +631,231 @@ def apply(self, *args, **kwargs):
+        code. By default true.
+        WARNING: there is something strange in the
+        behaviour of the slow implementation:
+        Only outputs, used outside this class, can be found using the
```
That's probably because they are in the inner graph, which by the way can be accessed using `scan_variables` attribute of `ComputationGraph`.

rizar commented on blocks/bricks/recurrent.py:
```diff
@@ -630,3 +631,231 @@ def apply(self, *args, **kwargs):
+        other layers and is added to the
+        usually input comfing from a fork from the states of the previous
+        layer.
+        For this to work it is assumed that all layers have the same names
```
Such assumptions should be checked in the code.

rizar commented on blocks/bricks/recurrent.py:
```diff
@@ -630,3 +631,231 @@ def apply(self, *args, **kwargs):
+                for s in getattr(transition.apply, t):
+                    layer_kwargs[s] = kwargs.get(s + suffix)
+
+            for k in ['iterate', 'reverse', 'return_initial_states']:
```
I am not a big fan of hard-coding here keyword arguments `recurrent_apply` can accept.

udibr replied:
Yes it is bad but this method (do_apply) in its fast_apply mode is mimicking the behaviour of recurrent_apply so it should be aware of its argumenrs somehow (any other idea will be welcomed.)

rizar commented on blocks/bricks/recurrent.py:
```diff
@@ -630,3 +631,231 @@ def apply(self, *args, **kwargs):
+            sequences_names = transition.apply.sequences
+            if d == 0:
+                layer0_kwargs = dict(
+                    (s, kwargs.get(s)) for s in sequences_names)
```
I it is future-proof to assume that the only arguments a recurrent application method takes are sequences. You should pass all arguments, that i.s. `layer0_kwargs = kwargs`

udibr replied:
OK maybe this is the solution ... looking at it now...

rizar replied:
You can't handle all of them in this way though. For instance `return_initial_states=True` requires special handling because from the initial states inputs should not be forked.

udibr replied:
There is a special if for return_initial_states few lines below this

udibr commented:
Took all your comments into account. As for speed, I noticed that the main limit is not from the GPU which is only ~70% utilized but from my CPU. On the CPU side only a single process (core) is active at 100%. My code does not do anything else in addition to RNN. However, it could be that I am using an external GPU on a Mac that causes the CPU to work so hard.

rizar commented on blocks/bricks/recurrent.py:
```diff
@@ -630,3 +631,241 @@ def apply(self, *args, **kwargs):
+        # Programmatically set the apply method
+        self.apply = self.fast_apply if fast else self.low_memory_apply
+        setattr(self.apply, 'sequences', transitions[0].apply.sequences)
+        for t in ["states", "contexts", "outputs"]:
```
This piece of code needs refactoring. We try not to use abbreviations (except very common ones) and one character identifies in Blocks. You can use `level` instead of `d`, `name` instead of `s`, `property_` instead of `t`, etc.

rizar commented on blocks/bricks/recurrent.py:
```diff
@@ -630,3 +631,241 @@ def apply(self, *args, **kwargs):
+
+        # sum up all the arguments we exepct to see in a call to a transition
+        # apply method, anything else is a recursion control
+        self.special_args = set(sum(map(lambda t: getattr(self.apply, t),
```
I think `set(self.apply.sequences + ... + ['mask'])` will be much simpler for the user to read.

rizar commented on blocks/bricks/recurrent.py:
```diff
@@ -630,3 +631,241 @@ def apply(self, *args, **kwargs):
+        """
+        results = []
+        last_states = None
+        layer0_kwargs = None
```
You can already initialize `layer0_kwargs` here, there is no need to do it inside the loop

rizar commented on blocks/bricks/recurrent.py:
```diff
@@ -630,3 +631,241 @@ def apply(self, *args, **kwargs):
+                    layer_kwargs["mask"] = kwargs.get("mask")
+
+            # handle all arguments that should be made unique to each layer
+            for t in ["states", "contexts", "outputs"]:
```
Also here, `s` and `t` are quite bad for readability.

rizar commented on blocks/bricks/recurrent.py:
```diff
@@ -630,3 +631,241 @@ def apply(self, *args, **kwargs):
+                layer_kwargs[k] = kwargs[k]
+
+            result = transition.apply(as_list=True, **layer_kwargs)
+
```
I think this blank line is a bit too much, but this is up to you.

rizar commented:
Cool, it is getting better and better! Some work on readability is still necessary, see my comments in the code.

I would also like to suggest you a bit different design for skip connections. I think it would be better to move the transformations that yield them outside the brick. Instead, you can add additional inputs to each of the layers. That is, if `skip_connections` is `False`, only first layer gets inputs, and if `skip_connections=True`, all the layers get inputs, which are summed with the contributions from the lower states.

udibr commented:
The idea to put the skip inputs outside will allow for more flexibility but I see some problems with it:
1) It will complicate the default usage of this brick. Right now it is almost transparent to add it to a single layer network (you just have to decide which states_... you want to take out instead of the single states of the single layer network.)
2) If I understand you correctly, with skip_connect=True I will have multiple sequence arguments in the input with a suffix number. For examples, inputs_0, inputs_1, ... but they are not fully symetric for now because inputs_0 needs a bias while inputs_1, inputs_2,... dont need a bias because there is already a bias term coming from the fork_1, fork_2, ...
Perhaps the best solution for this would have been to move the bias to be inside the LSTM (or whatever other unit you are using.) In this case I could have used the default fork also for a single layer network.
Today when using SequenceGenerator there is a small problem that the defsult fork that feeds the input to the LSTM does not have a bias and I manually have to replace it with a fork with bias...
3) the inputs_0, inputs_1, ... are not symmetric also in respect to the fact that inputs_0 is mandaroy while _1,... are not. Perhaps the name of inputs_0 should be kept just "inputs" while all other keep their number... starting to be confusing...

rizar commented:
I created an issue about bad default prototype for `Fork`, #703 

Answering your concern (1): typical usage of all recurrent bricks is in a combination with `Fork`. If it were enough to create a `Fork` that would fork inputs for all levels, usage would be very similar to a usual recurrent brick. But as you rightly pointed out, the first layer inputs need biases and the others don't. That is if we implement `skip_connections` by adding more inputs, we will have a recurrent transition for which the correct strategy of forking multiple inputs from a single common one less trivial than usually.

In fact I have been thinking that in addition to these core transition bricks like `SimpleRecurrent`, `GatedRecurrent` and `LSTM` we should also have similar ones with embedded forking (I guess this corresponds to your proposal "perhaps the best solution for this would have been to move the bias to be inside the LSTM"). This would also the nasty issue #586, and in fact in my dirty fork I already have the following dirty prototype: https://gist.github.com/rizar/36451d08719833fb59d0

If we go this way, the current `LSTM` will be become e.g. `CoreLSTM` and so on. Similarly there can be two recurrent stacks: `CoreRecurrentStack`, which takes many inputs, and `RecurrentStack`, which only takes one and takes care of forking it correctly, with or without biases depending on whether its bottom-most input or a skip-connection one.

Regarding (3), inputs can be different depending on `skip_connections=True/False` since you set them programatically.

What do you think about `RecurrentStack` and `CoreRecurrentStack`? The former can inherit from the latter one, and this way no so many change in your current code would be required.

udibr commented:
I think the CoreRecurrentStack idea is good.
What I would do is say that when calling its apply method then all inputs are mandatory
and that all the input names have a suffix (as is already the case with states and contexts)
In addition in CoreRecurrentStack the fork between the layers will not have a bias (today it has).

The wrapper RecurrentStack will add _0 to all inputs.
The skip_connections is an argument to the wrapper (not to the core.)
If skip_connections is True then the wrapper will add a fork of the inputs
to each of the other layers but this fork will have bias (today it does not.)
If the skip_connections is False the wrapper will add a new type of brick for each of the other layer.
This new brick will be a Bias class that  does not have any input (I think you can write something like that)
Either way, the inputs will compensate for the missing bias from layer to layer inside the core.

The wrapper will be used exactly as it is today, but, the names assigned to the bias will be different
as a result you cant re-load an old model to it....

Since the interface is the same, I think we can wait in implementing the core until you decide what to do with bias in general...

udibr commented:
I reallized that the contexts should be the same for all layers and therefore should not have a suffix with layer number. I'm working on a fix and also trying out the separation to Core you suggested.

rizar commented:
Good point about contexts.

One thing I find inconsistent in your version of `CoreRecurrentStack` is that without `RecurrentStack` it will be entirely useless,  because there is no biases between layers. On the contrary the usual "core" transition are in a sense complete and can be used on their own.

I think I know how to make `CoreRecurrentStack` very similar to the basic recurrent bricks we have now. If `skip_connections=True`, `Linear(use_bias=False)` should be used as `fork_prototype` and the `apply.sequences` should contain inputs for all layers with layer indices for disambiguation. On the other hand, if `skip_connections=False`, `Linear(use_bias=True)` should be the fork and only `transitions[0].sequences` should be in `apply.sequences`. This way all inputs will need biases, just as for the other current recurrent bricks. What do you think?

udibr commented:
My current code has the idea that skip_connections is an init parameter only of RecurrentStack and not of CoreRecurrentStack. This is because CoreRecurrentStack is "simple" and always expect to see inputs (ie sequences) to all layers.

However, you have a point that CoreRecurrentStack should be self contained by default. So what I can do is make the proto for the fork between layers have bias by default set to True.
This will make CoreRecurrentStack work by default if you only supply it inputs to layer 0.
In other words all your input arguments should be suffixed with _0.
(I can even add code that detects input names without a suffix and renames them.)

If someone wants to use CoreRecurrentStack directly and also supply input to other layers he will have to take this into account or replace the proto type for the fork.

This also solves me from the crazy Bias brick I was talking about (very easy to implement but ugly.)

What do you think?

rizar commented:
But your current code does not have two separate classes currently, it only has `RecurrentStack`. 

I am not sure that `skip_connections` should be introduced at the `RecurrentStack` level. What I had mind for `XXX` and `CoreXXX` classes is that the former just adds forking to the latter one. 

Also I think it would be nicer to keep the input names without suffixes for the case when skip connections are not used. My consideration is that in a well-designed software the one not using the feature (`skip_connections`) should not pay for its existence, and without this feature it would be reasonable to simply use `transition[0].apply.sequences` as `apply.sequences`.

udibr commented:
After walking in circules for few days I came back to where I started - almost.
I tried to split the code to core and wrapper and it turned out at the end that the wrapper did not contributed significantly to reduce the complexity or improve the functionality.
As a result of all the code shakeup, I fixed few bugs.
I changed the fork between the layers to use bias, as a result the class can be used out-of-the box.
I also made the names that appear in the apply parameters of that are comming from the bottom layer (layer 0) to be unmodified. As a result all the input interface to the apply became much more simpler to someone who does not want to insert inputs to other layers.
From runs I am doing now on https://github.com/udibr/sketch it looks like the RNN behaves better than before so probably I did fix some important bugs.

rizar commented on blocks/bricks/recurrent.py:
```diff
@@ -630,3 +631,281 @@ def apply(self, *args, **kwargs):
+        self.apply = self.fast_apply if fast else self.low_memory_apply
+        for level, transition in enumerate(transitions):
+            for property_ in ["sequences", "states", "outputs"]:
+                v = getattr(self.apply, property_, []) +\
```
Please use something a bit more verbose than `v`, e.g. `new_value`.

rizar commented on blocks/bricks/recurrent.py:
```diff
@@ -630,3 +631,281 @@ def apply(self, *args, **kwargs):
+
+        if fork_prototype is None:
+            # By default use Linear (with bias) as a default fork.
+            fork_prototype = Linear(use_bias=True)
```
If we write here

```python
fork_prototype = Linear(use_bias=not skip_connections)
```

then I think all the `skip_connections` bricks will be not necessary anymore. Everything can be done by an external fork _with_ biases.

rizar commented:
The code actually has become much better, but I am still in favour of moving things out of this brick, see my comments in the code.

udibr commented:
I finally understood what you wanted (sorry for being so slow.)
You want the external single fork (for example the fork inside SequenceGenerator) to 
generate all inputs for all layers when skip_connections is True.

1) I fixed the code, hoepfully it is what you wanted.
2) I fixed the docstring so someone else reading it will have an easier time understanding all this. (please read)
3) I wounder if blocks in general is not missing the point if it is so easy to miss understand how to use it.

rizar commented on blocks/bricks/recurrent.py:
```diff
@@ -630,3 +631,295 @@ def apply(self, *args, **kwargs):
+class RecurrentStack(BaseRecurrent, Initializable):
+    u"""Stack of recurrent networks.
+
+    Build a stack of recurrent layers from a supplied list of
```
_builds_

rizar commented on blocks/bricks/recurrent.py:
```diff
@@ -630,3 +631,295 @@ def apply(self, *args, **kwargs):
+    u"""Stack of recurrent networks.
+
+    Build a stack of recurrent layers from a supplied list of
+    BaseRecurrent objects. Each object must have a `sequences`,
```
:class:`~blocks.bricks.recurrent.BaseRecurrent`

rizar commented on blocks/bricks/recurrent.py:
```diff
@@ -630,3 +631,295 @@ def apply(self, *args, **kwargs):
+    such as the ones required by the recurrent decorator from
+    :mod:`blocks.bricks.recurrent`.
+
+    The apply parameters from all layers are concatenated.
```
"apply arguments" sounds a bit better, because the word "parameters" in Blocks already has certain semantics

rizar commented on blocks/bricks/recurrent.py:
```diff
@@ -630,3 +631,295 @@ def apply(self, *args, **kwargs):
+        the states of each layer. The transformation is used when state
+        is forked to the sequences of the next layer. By default a
+         :class:`.Linear` transformation is used, with bias if
+         skip_connections is False.
```
``False``

rizar commented on blocks/bricks/recurrent.py:
```diff
@@ -630,3 +631,295 @@ def apply(self, *args, **kwargs):
+         skip_connections is False.
+    states_name : string
+        In a stack of RNN the state of each layer is used as input to the
+        next. The `states_name` identify the element of the states of each
```
\'states\'

rizar commented on blocks/bricks/recurrent.py:
```diff
@@ -630,3 +631,295 @@ def apply(self, *args, **kwargs):
+        next. The `states_name` identify the element of the states of each
+        layer that should be used for this task. By default the element is
+        called "states". To be more precise, this is the name of the
+        element in the outputs of the apply method of each transition
```
\'outputs\' and also below \'sequences\'

rizar commented on blocks/bricks/recurrent.py:
```diff
@@ -630,3 +631,295 @@ def apply(self, *args, **kwargs):
+    fast : bool
+        Use the fast, but also memory consuming, implementation of this
+        code. By default true.
+    skip_connections : bool
```
A few words are missing to explain how it actually changes the model.

rizar commented on blocks/bricks/recurrent.py:
```diff
@@ -630,3 +631,295 @@ def apply(self, *args, **kwargs):
+        # Programmatically set the apply method
+        self.apply = self.fast_apply if fast else self.low_memory_apply
+        # parameters of base level are exposed as is
+        # excpet for mask which we will put at the ver end. See below.
```
_expect_, _very_

rizar commented on blocks/bricks/recurrent.py:
```diff
@@ -630,3 +631,295 @@ def apply(self, *args, **kwargs):
+
+        # add parameters of other layers
+        if skip_connections:
+            exposed_parameters = ["sequences", "states", "outputs"]
```
"arguments" would be better, because parameters are in 99.99% shared variables

rizar commented on blocks/bricks/recurrent.py:
```diff
@@ -630,3 +631,295 @@ def apply(self, *args, **kwargs):
+            sum([transition.apply.contexts for transition in transitions], [])
+        ))
+
+        # sum up all the arguments we exepct to see in a call to a transition
```
_expect_

rizar commented on blocks/bricks/recurrent.py:
```diff
@@ -630,3 +631,295 @@ def apply(self, *args, **kwargs):
+        All return values are of type :class:`~tensor.TensorVariable`.
+
+        """
+        nargs = len(args)
```
It would be cool to support positional arguments here, but then contexts should also be supported. For the first version supporting only keyword arguments would be sufficient I would say.

rizar replied:
This has not yet been addressed.

rizar commented on blocks/bricks/recurrent.py:
```diff
@@ -630,3 +631,295 @@ def apply(self, *args, **kwargs):
+            if level > 0:
+                # add the forked states of the layer below
+                inputs = self.forks[level - 1].apply(last_states, as_list=True)
+                for name, input in zip(normal_inputs, inputs):
```
Please use `input_` to avoid this crazy highlighting.

rizar commented on tests/bricks/test_recurrent.py:
```diff
@@ -205,6 +205,197 @@ def sigmoid(x):
+        depth = 4
+        self.depth = depth
+        dim = 3  # don't change, hardwired in the code
+        tarnsitions = [LSTM(dim=dim) for _ in range(depth)]
```
_transitions_

rizar commented:
Perfect, this is exactly what I wanted! Hopefully it makes this new class almost interchangeable with all other recurrent bricks.

This is one more (hopefully) serious concern: `reverse=True` will break the current code because it will be applied as many times as many layers you have. Without skip connections a fix would be to pass this option only to the bottom-most layer.  With skip connections things become more involved, unfortunately. Seems like explicitly reversing the inputs would be a solution.

I am looking forward to merge this PR this week. I will be however unavailable from now until Thursday due to relocation.

Regarding understanding Blocks: point taken, we definitely need more tutorials, better documented examples, etc..  I should have time this month to sit down and at least finally compose a brick development tutorial, which explain what is a brick, why we decided to support multiple application methods, what are properties of application methods, why they are necessary and how to set them.

udibr commented:
Fixed everything and add more comments on usage.
I can add an assert that reverse must be False...

rizar commented:
> I can add an assert that reverse must be False...

OK, this will do the job for now.

rizar commented:
A clarification: it should be not `assert` but an exception with a sensible error message.

rizar commented on blocks/bricks/recurrent.py:
```diff
@@ -630,3 +631,331 @@ def apply(self, *args, **kwargs):
+        kwargs.update(zip(self.apply.sequences[:nargs], args))
+
+        if kwargs.get("reverse", False):
+            raise Exception("reverse=True not implemented.")
```
That one should be `NotImplementedError` I guess.

rizar commented on blocks/bricks/recurrent.py:
```diff
@@ -630,3 +631,331 @@ def apply(self, *args, **kwargs):
+        return tuple(results)
+
+    @recurrent
+    def low_memory_apply(self, *args, **kwargs):
```
I am worried that the way you provide the choice between the fast and the memory efficient one is rather quirky and that the `apply` method ends up having no docs at all. Alternatively one can go this way:

```
def do_apply(self, *args, **kwargs):
    pass

@recurrent
def low_memory_apply(self, *args, **kwargs):
    return self.do_apply(*args, **kwargs)

@application
def apply(self, *args, **kwargs):
    # all docs go here!
    if kwargs.get('low_memory', False):
        return self.low_memory_apply(*args, **kwargs)
    return self.do_apply(*args, **kwargs)
```

As a consequence the `fast` flag would be not required.

rizar commented:
There are only a few issues left, after they are fixed this PR can be merged.

EloiZ commented:
Hi,
I am really interested in this PR. I have already started using it, however, I cannot figure out why the `fast_apply` method is faster than `low_memory_apply`. I thought that one scan wrapping the whole thing (`low_memory_apply`) would be faster than one scan for each layer (`fast_apply`)

udibr commented:
I've added the @apply as you suggested

rizar commented:
All right, thanks! There is still one unaddressed comment in the code, but I will merge. Great job!

rizar commented:
Created #724
